agent/lib_composer.c

```diff
@@ -0,0 +1,233 @@
+}
+
+static void nr_execute_handle_autoload_composer_get_packages_information(
+    const char* vendor_path) {
+  zval retval;  // This is used as a return value for zend_eval_string.
+                // It will only be set if the result of the eval is SUCCESS.
+  int result = -1;
+
+  // nurunlikely because this should alredy be ensured by the caller
+  if (nrunlikely(!NRINI(vulnerability_management_package_detection_enabled))) {
+    // do nothing when collecting package information for vulnerability
+    // management is disabled
+    return;
+  }
+
+  // nurunlikely because this should alredy be ensured by the caller
+  if (nrunlikely(!NRINI(vulnerability_management_composer_api_enabled))) {
+    // do nothing when use of composer to collect package info is disabled
+    return;
+  }
+
+  // clang-format off
+  char* getallrawdata
+        = ""
+        "(function() {"
+        "  try {"
+        "    $root_package = \\Composer\\InstalledVersions::getRootPackage();"
+        "    $packages = array();"
+        "    foreach (\\Composer\\InstalledVersions::getAllRawData() as $installed) { "
+        "      foreach ($installed['versions'] as $packageName => $packageData) {"
+        "        if (is_array($root_package) && array_key_exists('name', $root_package) && $packageName == $root_package['name']) {"
+        "          continue;"
+        "        }"
+        "        if (isset($packageData['pretty_version'])) {"
+        "          $packages[$packageName] = ltrim($packageData['pretty_version'], 'v');"
+        "        }"
+        "      }"
+        "    }"
+        "    return $packages;"
+        "  } catch (Exception $e) {"
+        "    return NULL;"
+        "  }"
+        "})();";
+  // clang-format on
+
+  if (NR_SUCCESS != nr_execute_handle_autoload_composer_init(vendor_path)) {
+    nrl_debug(NRL_INSTRUMENT,
+              "%s - unable to initialize Composer runtime API - package info "
+              "unavailable",
+              __func__);
+    return;
+  }
+
+  nrl_verbosedebug(NRL_INSTRUMENT, "%s - Composer runtime API available",
+                   __func__);
+
+  result
+      = zend_eval_string(getallrawdata, &retval, "composer_getallrawdata.php");
+  if (SUCCESS != result) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s - composer_getallrawdata.php failed",
+                     __func__);
+    return;
+  }
+  if (IS_ARRAY == Z_TYPE(retval)) {
+    zend_string* package_name = NULL;
+    zval* package_version = NULL;
+    ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL(retval), package_name,
+                                  package_version) {
+      if (NULL == package_name || NULL == package_version) {
+        continue;
+      }
+      if (nr_php_is_zval_non_empty_string(package_version)) {
+        nrl_verbosedebug(NRL_INSTRUMENT, "package %s, version %s",
+                         NRSAFESTR(ZSTR_VAL(package_name)),
+                         NRSAFESTR(Z_STRVAL_P(package_version)));
+        nr_txn_add_php_package_from_source(
```

Comment:
Doesn't NRSAFESTR return the string `"<NULL>"` when it encounters a null?
So if using NRSAFESTR with the call to nr_txn_add_php_package_from_source, a null package_name/package_version will actually send valid strings (that say NULL) and won't hit the check for a null package_name [here](https://github.com/newrelic/newrelic-php-agent/pull/962/files#diff-aff431af9f447c5431ea6466c518a37c44b2491884a829f6c9e4675082ca8186R3509) or [here](https://github.com/newrelic/newrelic-php-agent/pull/962/files#diff-bffd93c552e94ddfafc8445c041ff9e4727959267c975fa3c62f67c185561df7R40)
And it'll miss the check for a null version [here](https://github.com/newrelic/newrelic-php-agent/pull/962/files#diff-bffd93c552e94ddfafc8445c041ff9e4727959267c975fa3c62f67c185561df7R50)

Reply:
@zsistla `package_name` and `package_version` are guaranteed to not be `NULL` [here](https://github.com/newrelic/newrelic-php-agent/blob/cb83165d367509f3bb3439ad14b0aa61a21225c3/agent/lib_composer.c#L137). Additionally, `package_version` is guaranteed to be a non-empty string [here](https://github.com/newrelic/newrelic-php-agent/blob/cb83165d367509f3bb3439ad14b0aa61a21225c3/agent/lib_composer.c#L140). I don't see a way for either `package_name` or `package_version` ending up as `"<NULL>"` generated by use of `NRSAFESTR`. To be frank, because of those earlier `NULL` checks, `NRSAFESTR` could be dropped altogether.

Reply:
Sounds appropriate to drop NRSAFESTR when passed to nr_txn_add_php_package_from_source, it didn't seem the right way to use it there, but if you remove it from nr_txn_add_php_package_from_source, I'm good with it.

Reply:
Dropped `NRSAFESTR` in 3a356ea74f4fdc5cb74c09589fd98248b4d8bc2f.

Reply:
NRSAFESTR was okay (and safer to leave) when passing to nrl_*.
I only objected to it when passing to nr_txn_add_php_package_from_source

Reply:
The [same argumentation](https://github.com/newrelic/newrelic-php-agent/pull/962#discussion_r1773696404) applies when `NRSAFESTR` is used in `nrl_verbosedebug`. Let's be consistent.

Reply:
I don't believe the same argument applies.  nrl_verbosedebug does not handle NULLs so needs to be passed a safestr where NULL has been converted to  the string "<NULL>" to display and would segfault if passed an actual NULL.
whereas nr_txn_add_php_package_from_source does handle NULLs fine and would in fact behave incorrectly if the string "<NULL>" were passed to it instead of an actual NULL.  
If we dealing with anything to do with zvals (which are technically shared space values that PHP or another entity could muck with), it's better to be careful (the reason why we usually strdup anything quickly that we've gotten from a zval).

Reply:
`package_name` and `package_version` are guaranteed to not be NULL [here](https://github.com/newrelic/newrelic-php-agent/blob/cb83165d367509f3bb3439ad14b0aa61a21225c3/agent/lib_composer.c#L137). Additionally, `package_version` is **guaranteed to be a non-empty string** [here](https://github.com/newrelic/newrelic-php-agent/blob/cb83165d367509f3bb3439ad14b0aa61a21225c3/agent/lib_composer.c#L140) and therefore use of `NRSAFESTR` could be dropped no matter where the value of at least `package_version` is accessed. However, as requested I restored use of `NRSAFESTR` when `package_name` and `package_version` are accessed for debug log message in 51f6d43a7dd9a41b4466ea6e9dd3796c6af9c656. Maybe a better approach would be to guarantee that `package_name` is a non-empty string with `nr_php_is_zval_non_empty_string` to strengthen [this check](https://github.com/newrelic/newrelic-php-agent/blob/cb83165d367509f3bb3439ad14b0aa61a21225c3/agent/lib_composer.c#L137).